packages/yarnpkg-fslib/sources/ZipOpenFS.ts

```diff
@@ -247,16 +247,16 @@ export class ZipOpenFS extends BasePortableFakeFS {
   async realpathPromise(p: PortablePath) {
     return await this.makeCallPromise(p, async () => {
       return await this.baseFs.realpathPromise(p);
-    }, async (zipFs, {archivePath, subPath}) => {
-      return this.pathUtils.join(await this.baseFs.realpathPromise(archivePath), this.pathUtils.relative(PortablePath.root, await zipFs.realpathPromise(subPath)));
+    }, async (zipFs, {subPath}) => {
+      return this.pathUtils.join(zipFs.getRealPath(), this.pathUtils.relative(PortablePath.root, await zipFs.realpathPromise(subPath)));
```

Comment:
This should be fine as we've already ran the path through the `baseFs` in `findZip` here https://github.com/yarnpkg/berry/blob/7fcd174e610c3ebdd60391f1b915eff7f7f51431/packages/yarnpkg-fslib/sources/ZipOpenFS.ts#L789

Also they technically don't do the same thing `zipFs.getRealPath()` can return a symlink so we'd need a `zipFs.getRealRealPath()` or make it actually return the real path or just ignore it :thinking:

Reply:
I'm not sure I follow - `getRealPath` will return the physical path when called on JailFS, so calling `realpath` on a `ZipOpenFS(JailFS())` would return the physical path - which would then be interpreted by other functions as relative to the jail root, which would be breaking.

I'm not sure we can semantically avoid the existing implementation of this function.

Reply:
> I'm not sure I follow

I thought you meant that `ZipFs.getRealPath()` could escape the jail, but that path has already been through `lstatSync` but I misunderstood

> getRealPath will return the physical path when called on JailFS, so calling realpath on a ZipOpenFS(JailFS()) would return the physical path - which would then be interpreted by other functions as relative to the jail root, which would be breaking.

Ah, yeah that's true. I'll use the `archivePath` and cache the `realpath` result for as long as the `ZipFS` instance is alive then

Reply:
Should be fine to add a `realArchivePath` parameter to the `makeCall` callback 👍 

As for cache lifetime, I think it's fine to cache it forever since it's already what we do for `isZip` / `notZip`. At some point maybe we'll have to think about cache invalidation by trapping some calls, but no need to overdo it for now.

Reply:
> Should be fine to add a realArchivePath parameter to the makeCall callback

That would require it to always be computed so I instead made `realpath(Promise|Sync)` populate the cache as needed